### PR TITLE
Slow test result handler

### LINKED ISSUE #90: Show ten slowest tests report
it would be useful for haas to show the 10 slowest tests in the summary report.

### PR DESCRIPTION
**Pull request by sjagoe**
@itziakos Can you have a look over this and let me know what you think?

Closes #90

### CODE REVIEW

**Comment by sjagoe** [comment]
Example output:

```
$ haas --summarize-slow-tests 
................................................................................................................................................................................................
----------------------------------------------------------------------
Ran 192 tests in 0.191s

OK

----------------------------------------------------------------------
  0:00:0.013 test_main_buffer (haas.tests.test_haas_application.TestHaasApplication)
  0:00:0.009 test_with_coverage_plugin (haas.tests.test_haas_application.TestHaasApplication)
  0:00:0.006 test_main_default_arguments (haas.tests.test_haas_application.TestHaasApplication)
  0:00:0.006 test_main_failfast (haas.tests.test_haas_application.TestHaasApplication)
  0:00:0.005 test_multiple_start_directories (haas.tests.test_haas_application.TestHaasApplication)
  0:00:0.005 test_with_logging (haas.tests.test_haas_application.TestHaasApplication)
  0:00:0.005 test_main_quiet (haas.tests.test_haas_application.TestHaasApplication)
  0:00:0.004 test_multiple_start_directories_non_package (haas.tests.test_haas_application.TestHaasApplication)
  0:00:0.004 test_main_verbose (haas.tests.test_haas_application.TestHaasApplication)
  0:00:0.004 test_discover_creates_importerror_testcase (haas.plugins.tests.test_discoverer.TestDiscovererImportError)
```

**Comment by itziakos** [comment]
(ui) Can we add a heading on the output like `10 slowest tests`

File: haas/plugins/result_handler.py [hunk 1]
```diff
@@ -13,6 +13,14 @@
 from .i_result_handler_plugin import IResultHandlerPlugin
 
 
+def get_test_description(test, descriptions=True):
+    doc_first_line = test.shortDescription()
+    if descriptions and doc_first_line:
+        return '\n'.join((str(test), doc_first_line))
```
Comment (itziakos) [inline haas/plugins/result_handler.py]: Should this be `unicode`?

File: haas/plugins/result_handler.py [hunk 2]
```diff
@@ -217,3 +221,63 @@ def __call__(self, result):
...
+    def print_summary(self):
+        tests_by_time = sorted(
+            self._test_results,
+            key=lambda item: item.timing,
+            reverse=True,
+        )
+
+        self.stream.writeln()
+        self.stream.writeln(self.separator2)
+
```
Comment (itziakos) [inline haas/plugins/result_handler.py]: how about write#ing a header like
```
self.stream.writeln('{} slowest tests'.format(self.number_to_summarize))
self.stream.writeln(self.separator2)
```

File: haas/result.py [hunk 3]
```diff
@@ -98,6 +98,63 @@ def _format_exception(err, is_failure, stdout=None, stderr=None):
     return ''.join(msgLines)
 
 
+class TestTiming(object):
+
```
Comment (itziakos) [inline haas/result.py]: (nitpick) docstring

**Comment by itziakos** [comment]
minor comments, I am very happy with this :+1:

File: haas/plugins/result_handler.py (hunk 1 above)
Comment (sjagoe) [inline haas/plugins/result_handler.py]: I'll need to re-confirm behaviour with `unittest`, but in this PR the code has just been moved unchanged from a method to a function to be re-used elsewhere.

**Comment by sjagoe** [comment]
I'll need to work on testing; this was failed by coveralls :-(

**Comment by sjagoe** [comment]
The order of test result handlers is not consistent.

File: haas/plugins/result_handler.py [hunk 4]
```diff
@@ -217,3 +221,63 @@ def __call__(self, result):
             self.stream.write(" '{0}'".format(result.message))
         self.stream.writeln()
         self.stream.flush()
+
+
+class SlowTestsResultHandler(IResultHandlerPlugin):
+    separator1 = '=' * 70
+    separator2 = separator2
+
+    OPTION_DEFAULT = object()
```
Comment (itziakos) [inline haas/plugins/result_handler.py]: (suggestion) I would avoid mutable objects as class attributes.

File: haas/plugins/result_handler.py [hunk 5]
```diff
@@ -217,3 +221,63 @@ def __call__(self, result):
             self.stream.write(" '{0}'".format(result.message))
         self.stream.writeln()
         self.stream.flush()
+
+
+class SlowTestsResultHandler(IResultHandlerPlugin):
+    separator1 = '=' * 70
```
Comment (itziakos) [inline haas/plugins/result_handler.py]: Would it make sense to have these as optional arguements in ``__init__``

File: haas/plugins/result_handler.py (hunk 4 above)
Comment (sjagoe) [inline haas/plugins/result_handler.py]: This is not mutable:

```[nix-shell:~/workspace/source/haas]$ python
Python 3.4.4 (default, Dec 21 2015, 06:01:07) 
[GCC 5.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> o = object()
>>> o.foo = 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'object' object has no attribute 'foo'
>>> 
(haas34)
[nix-shell:~/workspace/source/haas]$ python2
Python 2.7.11 (default, Dec  5 2015, 19:47:20) 
[GCC 5.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> o = object()
>>> o.foo = 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'object' object has no attribute 'foo'
```

File: haas/plugins/result_handler.py (hunk 5 above)
Comment (sjagoe) [inline haas/plugins/result_handler.py]: For what reason?

Comment (itziakos) [inline haas/plugins/result_handler.py]: Just not very keen on class attributes. This is a very subjective thus can be ignored

Comment (sjagoe) [inline haas/plugins/result_handler.py]: Class attributes make sense when it is static data shared with all instances. There is no reason to put it in the instance `__dict__`.

**Comment by sjagoe** [comment]
Opened #149 for the plugin sort stability issue.

File: haas/plugins/result_handler.py (hunk 5 above)
Comment (itziakos) [inline haas/plugins/result_handler.py]: No problem

File: haas/plugins/result_handler.py (hunk 4 above)
Comment (itziakos) [inline haas/plugins/result_handler.py]: True, this always gets me (Apparently brain functions need food and coffee). Please ignore my comment

**Comment by itziakos** [comment]
:+1: can't wait to use this feature

**Pull request merged**